Proxy support

mrbaseman opened the pull request:
this branch implements the support proxy as requested in #150

DimitriPapadopoulos commented:
Can you fix the Codacy warning?

DimitriPapadopoulos commented:
@mrbaseman Part of the patch is not strictly following the [Linux kernel coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html), typically:
> ### 3) Placing Braces and Spaces
> [...]
> ### 3.1) Spaces
> [...]
> 
> Use one space around (on each side of) most binary and ternary operators, such as any of these:
> `=  +  -  <  >  *  /  %  |  &  ^  <=  >=  ==  !=  ?  :`

I don't want to be pedantic though - only use this comment if you feel it it useful. Alternatively commit it as is and, if you agree, I can change the code later, before a release for example.

DimitriPapadopoulos commented:
By the way there's a [_checkpatch_ Perl script](https://wintergreenworks.wordpress.com/2012/05/30/checkpatch-linux-coding-style-check/) to check patches and files in the Linux kernel tree.

adrienverge commented:
@DimitriPapadopoulos is right, let's try to use the same coding rules consistently across the code. For example I find `bytes_read=read(handle,&(request[curr_pos++]),2);` not very readable.

> By the way there's a checkpatch Perl script to check patches and files in the Linux kernel tree.

I didn't know this script. Maybe it could be used in our CI!

DimitriPapadopoulos commented:
@adrienverge Unfortunately I find this script way too verbose. I had to leave out lots of errors, some of which I'm really not interested in.

In any case:
* We need to fix the existing code first.
* We might want to disregard out a few errors types.

adrienverge commented:
@DimitriPapadopoulos Sure, no problem.

mrbaseman commented:
our `lint` check was already happy with the current patch, but `checkpatch.pl` finds a couple more problems. I'll correct them.

DimitriPapadopoulos commented on src/tunnel.c:
```diff
@@ -322,9 +326,22 @@ static int tcp_connect(struct tunnel *tunnel)
 		server.sin_addr.s_addr = inet_addr(proxy_host);
+		// if host is given as fqhn we have to do a dns lookup
+		if (server.sin_addr.s_addr == INADDR_NONE) {
+			host = gethostbyname(proxy_host);
```
While _gethostname()_ is perfect right now, please note for future reference that pull request #77 suggests switching to _getaddrinfo()_. It makes sense as [_gethostname()_ is obsolete and not part of POSIX anymore](https://github.com/adrienverge/openfortivpn/pull/77#issuecomment-335367313).

mrbaseman commented:
Yes, I'm aware of that, but on line [271](https://github.com/adrienverge/openfortivpn/blob/6a32acd5b62fe6c8fb832212dd48a59c27967851/src/tunnel.c#L271) of tunnel.c `gethostbyname` is used too. When adding ipv6 support in the future we have to change both places.

mrbaseman commented:
I'll check how #77 needs to be updated when #192 is landed and push an appropriate commit to that branch (or vice versa, depending on which pr gets merged first).

DimitriPapadopoulos commented:
@mrbaseman Looks good to me. You just need to update the man page after commit 28cd7f0 _support dns lookup for proxy and tolerate trailing slashes_:
```
It checks if one of the environment variables
B https_proxy HTTPS_PROXY all_proxy ALL_PROXY
is set which are supposed to contain a string of the format 
.br
.B http://[ip]:[port]
.br
where
.B [ip]
is the ip of the proxy server in dotted decimal notation and
.B [port]
```
Indeed one may now use a DNS name instead of an IPv4 address.

DimitriPapadopoulos commented:
Also in **tunnel.c** and unless I'm missing something you can change:
```
		proxy_str = strdup(env_proxy);
		env_proxy = proxy_str;
```
into:
```
		env_proxy = strdup(env_proxy);
```

DimitriPapadopoulos commented:
And finally according to [The Open Group Base Specifications Issue 6
IEEE Std 1003.1, 2004 Edition](http://pubs.opengroup.org/onlinepubs/009695399/functions/bzero.html) you should use _memset_() instead of _bzero_():
> ### APPLICATION USAGE
> 
> The [_memset_()](http://pubs.opengroup.org/onlinepubs/009695399/functions/memset.html) function is preferred over this function.
> 
> For maximum portability, it is recommended to replace the function call to _bzero_() as follows:
> 
> ```
> #define bzero(b,len) (memset((b), '\0', (len)), (void) 0)
> ```
> 
> ### FUTURE DIRECTIONS
> 
> This function may be withdrawn in a future version.
> 
> ### CHANGE HISTORY
> 
> First released in Issue 4, Version 2.
> 
> #### Issue 5
> Moved from X/OPEN UNIX extension to BASE.
> 
> #### Issue 6
> 
> This function is marked LEGACY.

I cannot find _bzero_() in [The Open Group Base Specifications Issue 7
IEEE Std 1003.1-2008, 2016 Edition](http://pubs.opengroup.org/onlinepubs/9699919799/).

DimitriPapadopoulos commented:
Also you've replaced `gethostbyname` with `getaddrinfo` in `AC_CHECK_FUNCS` (I had forgotten that) but you can also add `ntohs` next to `inet_ntoa`.

mrbaseman commented:
I have squashed and rebased this branch and updated the commit message now. @DimitriPapadopoulos and @adrienverge do you think we can merge it like this?

DimitriPapadopoulos commented:
Looks good to me.

Also proxy support seems complicated to me. Am I correct that there are different proxies out there with variations in the protocol? In any case, since it already works for two users, I suggest you merge this pull request so that it's available out there in the field and can be tested by more users.

adrienverge commented:
If it's good to both of you then it's good to me!

mrbaseman commented:
@DimitriPapadopoulos well, there are standard http proxies, proxies that require authentication in order to use them, socks proxies... This one is the simplest implementation (without authentication), but with a little bit of tolerance about line endings and http headers, where the proxy might not exactly follow the standard
